### ISSUE TITLE
Keyboard Events Not Transmitting Between Machines Using Universal Control with Karabiner

### ISSUE THREAD

**Issue by Drincann** [pqrs-org/Karabiner-Elements#3658]
I am using Universal Control to operate two machines, one is a **MacBook Pro M1 Pro** (referred to as Machine A) and the other is a **Mac Studio** (referred to as Machine B). **It seems that keyboard events triggered on one machine directly connected to a keyboard do not get transmitted to the Karabiner on the other machine**.

I used Karabiner-EventViewer to monitor the behavior on both machines, and they both have Karabiner installed.

When I type on Machine A (for the MacBook Pro, it's the built-in keyboard, and for the Mac Studio, it's a Bluetooth-connected keyboard), no matter if I focus on Machine A or B's screen (even if my input affects an application on Machine B), I can only see "Keyboard & pointing events" scrolling in the EventViewer of Karabiner on Machine A.

The same is true in reverse; this behavior is symmetrical on both machines.

Is this the expected behavior? I recall using Universal Control to control two MacBook Pros in conjunction with Karabiner, and everything worked fine.

I am experiencing an issue similar to issue #3260, which, as of my writing, has not received any responses.

Finally, I want to express my deep reliance on Karabiner-Elements for my daily workflow. This software is essential for me, to the point that my work efficiency is significantly impacted without it. I greatly appreciate the community and everyone who contributes to this project. Your efforts make a substantial difference in my daily professional life, and I am grateful for any assistance or insights you can provide regarding this issue.

**Comment by Drincann** [comment]
Hi everyone,

I wanted to share that I've successfully resolved an issue I encountered while using Hammerspoon in conjunction with Karabiner. My goal was to remap the Caps Lock key to an unused button, but I initially faced difficulties getting this remapping to work with Universal Control. Fortunately, I found a solution and it's now working perfectly.

For anyone experiencing similar issues, you might find this solution helpful: 

https://github.com/Hammerspoon/hammerspoon/issues/3512#issuecomment-1661977782

**Comment by sachinsid** [comment]
Hi @Drincann , Curious on your use-case above. Were you able to get the Caps Lock (acting as a HyperKey) working under Universal Control using the above workaround?

**Comment by Drincann** [comment]
@sachinsid

```lua
function key_remapping()
    -- remap capslock to F13:
    status = os.execute("hidutil property --set '{\"UserKeyMapping\":[{\"HIDKeyboardModifierMappingSrc\": 0x700000039, \"HIDKeyboardModifierMappingDst\": 0x700000068}]}'") 
    if not status then
        hs.dialog.blockAlert("Key remapping failed", "Check with:\nhidutil property --get UserKeyMapping")
    end
end
key_remapping()
```

This code allows you to remap the Caps Lock key to F13. After remapping, you can detect events from `hs.keycodes.map["F13"]` using `hs.eventtap.new(table, lambda)` without relying on external tools like Karabiner.

```lua
hs.eventtap.new({hs.eventtap.event.types.keyDown}, function(e)
    local currKey = e:getKeyCode()
    local capslock = hs.keycodes.map["F13"]
    if currKey == capslock then
        -- code here...
    end
end):start()
```

I believe the reason this method is effective is that Universal Control operates at a lower level than Karabiner. This results in the Caps Lock mapping event not being properly transmitted between the two machines using Universal Control. However, when I switched from Karabiner's mapping configuration to using hidutil, my keyboard events first reach the layer where hidutil operates, and then they enter Universal Control for event distribution, determining which machine should respond to the event.

**Comment by sachinsid** [comment]
Thanks for your prompt response....I'll give it a try. However with your suggested  approach, my physical Caps Lock key will remain unused (in favor of F13 or any other key I decide to map to) correct? Hard to get over the memory as I already have many keyboard shortcuts hardwired to the Hyper Key.

**Comment by Drincann** [comment]
There might be some misunderstanding, so let me clarify the entire background.

Since Caps Lock cannot be directly monitored in Hammerspoon, it's necessary to use Karabiner to map Caps Lock to a certain mouse key (which I'll refer to as 'mouse'). This allows me to listen for 'mouse' events in Hammerspoon, thereby implementing the Hyper Key.

In reality, my actual Hyper Key is Caps Lock, but in the Lua script, I operate with 'mouse'.

Later, I encountered an issue where the Hyper Key didn't work on another device while using Universal Control. To address this, I replaced Karabiner with `key_remapping()`, which works well under Universal Control.

My original script just added a call to `key_remapping()` and changed the code that was listening for mouse events to listen for F13 instead.

In summary, the role of `key_remapping` is to substitute for Karabiner's Caps Lock to some key mapping.

The key it maps to is used for monitoring in the script, not as the actual Hyper Key on the keyboard.

**Comment by mdbraber** [comment]
@Drincann to check: your solution is to use Hammerspoon to remap the Capslock key and not use Karabiner at all (so Universal Control is working again)?

**Comment by Drincann** [comment]
@mdbraber 

Yes, that's correct. I’m using Hammerspoon to run the remap via `hidutil` and to handle the key event.
